tests/http/clients/ws-data.c

```diff
@@ -184,73 +148,175 @@ static void websocket_close(CURL *curl)
-    res = send_binary(curl, send_buf, len);
-    if(res)
-      goto out;
-    res = recv_binary(curl, send_buf, len);
-    if(res) {
-      fprintf(stderr, "recv_data(len=%ld) -> %d\n", (long)len, res);
+    size_t nwritten, nread, slen = len, rlen = len;
+    char *sbuf = send_buf, *rbuf = recv_buf;
+
+    memset(recv_buf, 0, plen_max);
+    while(slen || rlen || scount || rcount) {
+      sblock = rblock = 1;
+      if(slen) {
+        r = curl_ws_send(curl, sbuf, slen, &nwritten, 0, CURLWS_BINARY);
+        sblock = (r == CURLE_AGAIN);
+        if(!r || (r == CURLE_AGAIN)) {
+          fprintf(stderr, "curl_ws_send(len=%ld) -> %d, %ld (%ld/%ld)\n",
+                  (long)slen, r, (long)nwritten,
+                  (long)(len - slen), (long)len);
+          sbuf += nwritten;
+          slen -= nwritten;
+        }
+        else
+          goto out;
+      }
+      if(!slen && scount) { /* go again? */
+        scount--;
+        sbuf = send_buf;
+        slen = len;
+      }
+
+      if(rlen) {
+        size_t max_recv = (64 * 1024);
+        r = curl_ws_recv(curl, rbuf, (rlen > max_recv) ? max_recv : rlen,
+                         &nread, &frame);
+        if(!r || (r == CURLE_AGAIN)) {
+          rblock = (r == CURLE_AGAIN);
+          fprintf(stderr, "curl_ws_recv(len=%ld) -> %d, %ld (%ld/%ld) \n",
+                  (long)rlen, r, (long)nread, (long)(len - rlen), (long)len);
+          if(!r) {
+            r = check_recv(frame, len - rlen, nread, len);
+            if(r)
+              goto out;
+          }
+          rbuf += nread;
+          rlen -= nread;
+        }
+        else
+          goto out;
+      }
+      if(!rlen && rcount) { /* go again? */
+        rcount--;
+        rbuf = recv_buf;
+        rlen = len;
+      }
+
+      if(rblock && sblock) {
+        fprintf(stderr, "EAGAIN, sleep, try again\n");
+  #ifdef _WIN32
+        Sleep(100);
+  #elif defined(__TANDEM)
+        /* NonStop only defines usleep when building for a threading model */
+  # if defined(_PUT_MODEL_) || defined(_KLT_MODEL_)
+        usleep(100*1000);
+  # else
+        PROCESS_DELAY_(100*1000);
+  # endif
+  #else
+        usleep(100*1000);
+  #endif
+      }
+    }
+
+    if(memcmp(send_buf, recv_buf, len)) {
+      fprintf(stderr, "recv_data: data differs\n");
+      dump("expected:", (unsigned char *)send_buf, len, 0);
+      dump("received:", (unsigned char *)recv_buf, len, 0);
+      r = CURLE_RECV_ERROR;
       goto out;
     }
   }
 
 out:
-  if(!res)
+  if(!r)
     websocket_close(curl);
   free(send_buf);
-  return res;
+  free(recv_buf);
+  return r;
+}
+
+static void usage(const char *msg)
+{
+  if(msg)
+    fprintf(stderr, "%s\n", msg);
+  fprintf(stderr,
+    "usage: [options] url\n"
+    "  -m number  minimum frame size\n"
+    "  -M number  maximum frame size\n"
+  );
 }
 
 #endif
 
 int main(int argc, char *argv[])
 {
-#ifndef CURL_DISABLE_WEBSOCKETS
+#if !defined(CURL_DISABLE_WEBSOCKETS) && !defined(_MSC_VER)
```

Comment:
Disabling this for Visual Studio builds due to a lack of getopt is not ideal. Here is an [alternative](https://github.com/iotivity/iotivity/blob/master/resource/c_common/windows/src/getopt.c) that may work for Visual Studio however it has an Apache license. @bagder can apache licensed files be used in testing?

Reply:
We don't run these tests on Windows anyway yet so it's not really an added limitation.

I think the server-side is the larger complication there.

But I believe Apache licensed files should be OK to use in tests - outside of product code.

Reply:
Ok I was not aware of that. This PR looks fine to me but I am not an expert in WebSockets so you may want to have a look before landing it.